hudi-client/hudi-spark-client/src/main/scala/org/apache/hudi/SparkFileFormatInternalRowReaderContext.scala

```diff
@@ -77,14 +78,18 @@ class SparkFileFormatInternalRowReaderContext(baseFileReader: Option[Partitioned
           }
         }).asInstanceOf[ClosableIterator[InternalRow]]
     } else {
-      if (baseFileReader.isEmpty) {
-        throw new IllegalArgumentException("Base file reader is missing when instantiating "
-          + "SparkFileFormatInternalRowReaderContext.");
+      val key = generateKey(dataSchema, requiredSchema)
+      if (!readerMaps.contains(key)) {
+        throw new IllegalStateException("schemas don't hash to a known reader")
       }
-      new CloseableInternalRowIterator(baseFileReader.get.apply(fileInfo))
+      new CloseableInternalRowIterator(readerMaps(key).apply(fileInfo))
     }
   }
 
+  private def generateKey(dataSchema: Schema, requestedSchema: Schema): Long = {
```

Comment:
Do we have any cases that  different readers use the same data schema and requested Schema?

Reply:
The only time it would matter is if there are pushdown filters. This is just a temporary hacky solution for now. We will eventually build the reader inside of getFileRecordIterator. We need to incorporate an abstraction for filters when we do that

Reply:
Hi Jon, i really feel that if you can split this  PR into smaller PRs, that would be much easier for reviewers to understand and easier for the CI to pass.